File: src/go/tree/98.validate-binary-search-tree.go
package main

// 98. Validate Binary Search Tree
// Function signature expected by the harness: func(*TreeNode) bool
// Implementation uses DFS with range constraints (open interval).
func isValidBST(root *TreeNode) bool {
	var dfs func(n *TreeNode, low *int, high *int) bool
	dfs = func(n *TreeNode, low *int, high *int) bool {
		if n == nil {
			return true
		}
		if low != nil && n.Val <= *low {
			return false
		}
		if high != nil && n.Val >= *high {
			return false
		}
		// Left subtree: all values must be < n.Val
		if !dfs(n.Left, low, &n.Val) {
			return false
		}
		// Right subtree: all values must be > n.Val
		if !dfs(n.Right, &n.Val, high) {
			return false
		}
		return true
	}
	return dfs(root, nil, nil)
}

File: src/java/tree/98.validate-binary-search-tree.java
class P98 {
    // @lc app=leetcode id=98 lang=java
    // @lc code=start
    public class Solution {
        public boolean isValidBST(TreeNode root) { return dfs(root, null, null); }
        private boolean dfs(TreeNode n, Long lo, Long hi){
            if (n==null) return true;
            long v = n.val;
            if ((lo!=null && v<=lo) || (hi!=null && v>=hi)) return false;
            return dfs(n.left, lo, v) && dfs(n.right, v, hi);
        }
    }
    // @lc code=end
}
